internal/oci/notation.go

```diff
@@ -0,0 +1,241 @@
+	return func(o *options) {
+		o.Auth = auth
+	}
+}
+
+// WithNotaryKeychain is a functional option for overriding the default
+// remote options used by the verifier
+func WithNotaryKeychain(key authn.Keychain) Options {
+	return func(o *options) {
+		o.Keychain = key
+	}
+}
+
+// NotatryVerifier is a struct which is responsible for executing verification logic
+type NotaryVerifier struct {
+	auth     authn.Authenticator
+	keychain authn.Keychain
+	verifier *notation.Verifier
+	opts     []remote.Option
+	insecure bool
+}
+
+type trustStore struct {
+	cert []byte
+}
+
+// GetCertificates implements truststore.X509TrustStore.
+func (s trustStore) GetCertificates(ctx context.Context, storeType truststore.Type, namedStore string) ([]*x509.Certificate, error) {
+	raw := s.cert
+	block, _ := pem.Decode(raw)
+	if block != nil {
+		raw = block.Bytes
+	}
+
+	certs, err := x509.ParseCertificates(raw)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse certificate '%s': %s", namedStore, err)
+	}
+
+	return certs, nil
+}
+
+// NewNotaryVerifier initializes a new NotaryVerifier
+func NewNotaryVerifier(opts ...Options) (*NotaryVerifier, error) {
+	o := options{}
+	for _, opt := range opts {
+		opt(&o)
+	}
+
+	store := &trustStore{
+		cert: o.PublicKey,
+	}
+
+	verifier, err := verifier.New(o.TrustStore, store, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	switch {
+	case o.Auth != nil && o.Keychain != nil:
+		return nil, fmt.Errorf("please provide either authn.Authenticator or authn.Keychain")
+	case o.Auth == nil:
+		o.Auth = authn.Anonymous
+	}
+
+	return &NotaryVerifier{
+		auth:     o.Auth,
+		keychain: o.Keychain,
+		verifier: &verifier,
+		opts:     o.ROpt,
+		insecure: o.Insecure,
+	}, nil
+}
+
+// Verify verifies the authenticity of the given ref OCI image.
+// It returns a boolean indicating if the verification was successful.
+// It returns an error if the verification fails, nil otherwise.
+func (v *NotaryVerifier) Verify(ctx context.Context, ref name.Reference) (bool, error) {
+	url := ref.Name()
+	remoteRepo, err := oras.NewRepository(url)
+	if err != nil {
+		return false, err
+	}
+	remoteRepo.PlainHTTP = v.insecure
+
+	repo := registry.NewRepository(remoteRepo)
+
+	credentialProvider := func(ctx context.Context, registry string) (oauth.Credential, error) {
+		return oauth.EmptyCredential, nil
+	}
+
+	auth := v.auth
+	if v.keychain != nil {
+		source := stringResource{url}
+
+		auth, err = v.keychain.Resolve(source)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	if auth != authn.Anonymous {
+		authConfig, err := auth.Authorization()
+		if err != nil {
+			return false, err
+		}
+
+		credentialProvider = func(ctx context.Context, registry string) (oauth.Credential, error) {
+			if authConfig.Username != "" || authConfig.Password != "" || authConfig.IdentityToken != "" {
+				return oauth.Credential{
+					Username:     authConfig.Username,
+					Password:     authConfig.Password,
+					RefreshToken: authConfig.IdentityToken,
```

Comment:
Why are we discarding the `authConfig.RegistryToken`?

Reply:
That was an oversight on my part. Resolved in [5eed49b](https://github.com/fluxcd/source-controller/pull/1075/commits/5eed49be7adc1e08f04b6124c6d9497b1d142986).